core/eolearn/core/eodata_io.py

```diff
@@ -0,0 +1,260 @@
+
+    features_to_save = ((FeatureIO(filesystem, path),
+                         eopatch[(ftype, fname)],
+                         FileFormat.NPY if ftype.is_raster() else FileFormat.PICKLE,
+                         compress_level) for ftype, fname, path in eopatch_features)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        executor.map(lambda params: params[0].save(*params[1:]), features_to_save)
+
+
+def load_eopatch(eopatch, filesystem, patch_location, features=..., lazy_loading=False):
+    """ A utility function used by EOPatch.load method
+    """
+    features = list(walk_filesystem(filesystem, patch_location, features))
+    loading_data = [FeatureIO(filesystem, path) for _, _, path in features]
+
+    if not lazy_loading:
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            loading_data = executor.map(lambda loader: loader.load(), loading_data)
+
+    for (ftype, fname, _), value in zip(features, loading_data):
+        eopatch[(ftype, fname)] = value
+
+    return eopatch
+
+
+def walk_filesystem(filesystem, patch_location, features=...):
+    """ Recursively reads a patch_location and returns yields tuples of (feature_type, feature_name, file_path)
+    """
+    existing_features = {ftype: {} for ftype in FeatureType}
+    for ftype, fname, path in walk_main_folder(filesystem, patch_location):
+        existing_features[ftype][fname] = path
+
+    returned_meta_features = set()
+    queried_features = set()
+    for ftype, fname in FeatureParser(features):
+        if fname is ... and not existing_features[ftype]:
+            continue
+
+        if ftype.is_meta():
+            if ftype in returned_meta_features:
+                continue
+            fname = ...
+            returned_meta_features.add(ftype)
+
+        elif ftype not in queried_features and (fname is ... or fname not in existing_features[ftype]):
+            queried_features.add(ftype)
+            if ... not in existing_features[ftype]:
+                raise IOError('There is not {} in saved EOPatch'.format(ftype))
+
+            for feature_name, path in walk_feature_type_folder(filesystem, existing_features[ftype][...]):
+                existing_features[ftype][feature_name] = path
+
+        if fname not in existing_features[ftype]:
+            raise IOError('Feature {} does not exist in saved EOPatch'.format((ftype, fname)))
+
+        if fname is ... and not ftype.is_meta():
+            for feature_name, path in existing_features[ftype].items():
+                if feature_name is not ...:
+                    yield ftype, feature_name, path
+        else:
+            yield ftype, fname, existing_features[ftype][fname]
+
+
+def walk_main_folder(filesystem, folder_path):
+    """ Walks the main EOPatch folders and yields tuples (feature type, feature name, path in filesystem)
+    """
+    for path in filesystem.listdir(folder_path):
+        raw_path = path.split('.')[0].strip('/')
```

Comment:
is this filesystem agnostic?

Reply:
Yes, every subclass of `fs.base.FS` object has to implement `listdir` method ([source](https://github.com/PyFilesystem/pyfilesystem2/blob/master/fs/base.py#L172)).